outbound connection sequence

myronkscott opened the pull request:
allow for late loading of network bytes.  clean outbound queues of canceled messages if outbound is slow.

myronkscott commented:
did a fresh push based on some of my own review

chrisdennis reviewed (commented):
I've gone through this and think I have a handle on what's going on, and I think it makes sense. What I'm wondering now is if the code is getting too complex, and if we can rework some of the OO aspects to make things a little easier to follow? Maybe we should have a chat later in the week and see if I can explain that a little more clearly?

myronkscott commented:
that's fine.  I'm working on some tests.  I can give a go to rework some of the layers.  I was trying to change as little as possible while trying to implement threading changes.

myronkscott commented:
I added a commit that separates the concerns that were bothering me.  I split TCAction based messages from WireProtocol messages.  They are used differently and have different stages.  I added the load phase to TCAction messages so that they can be late loaded and converted to bytes right before being sent on the wire rather than sitting in a queue.  I still need to add tests for the cancel functionality as well as the stages on both wire and action messages.  Happy to chat whenever to address any further concerns you might have.

myronkscott commented:
reworked based on wheat we talked about.  added testing.  Please have a look when you get a chance.

chrisdennis commented on tc-messaging/src/main/java/com/tc/net/protocol/transport/WireProtocolGroupMessageImpl.java:
```diff
@@ -49,115 +49,104 @@
+        hdr.computeChecksum();
+        WireProtocolMessage msg = new WireProtocolMessageImpl(this.sourceConnection, hdr, new TCByteBuffer[] {msgs.read(msgLen)});
+        messages.add(msg);
+      } catch (IOException ioe) {
```
This catch and ignore seems dangerous. What we probably should be doing is reworking TCByteBufferInputStream to only throw `EOFException` since we know IOException is impossible. That should probably then be caught here and rethrown as some kind of protocol error that will get handled sanely?

chrisdennis reviewed (changes requested):
Gave this a quick review... a handful of things stuck out to me.

myronkscott commented on common/src/main/java/com/tc/net/core/TCConnectionImpl.java:
```diff
@@ -104,7 +109,8 @@ final class TCConnectionImpl implements TCConnection, TCChannelReader, TCChannel
   private final AtomicLong                      totalWrite                  = new AtomicLong(0);
-  private final ArrayList<WriteContext>         writeContexts               = new ArrayList<WriteContext>();
+  private final Queue<WriteContext>     writeContexts               = new ConcurrentLinkedQueue<>();
+  private final Semaphore               writeContextControl         = new Semaphore(1);
```
i considered that but why?  It is not meant to be reentrant and I'm not a huge fan of the call structure of a ReentrantLock if not needed.

myronkscott replied:
Nevermind.  ReentrantLock is better, no interrupt exception.  I was thinking of read write lock for some reason.

myronkscott commented on common/src/main/java/com/tc/net/core/TCConnectionImpl.java:
```diff
@@ -85,7 +90,7 @@ final class TCConnectionImpl implements TCConnection, TCChannelReader, TCChannel
   private final boolean                         clientConnection;              
   private final AtomicBoolean                   transportEstablished        = new AtomicBoolean(false);
-  private final LinkedList<TCNetworkMessage>    writeMessages               = new LinkedList<TCNetworkMessage>();
+  private final BlockingQueue<TCNetworkMessage>    writeMessages               = new ArrayBlockingQueue<>(MSG_GROUPING_MAX_COUNT);
```
you mean using tc.properties?  I can set it to a fixed number but it is kind of nice to be able to adjust if desired.

myronkscott commented on common/src/main/java/com/tc/net/core/TCConnectionImpl.java:
```diff
@@ -402,50 +410,75 @@ private int doWriteToBuffer() throws IOException {
-        ? new ArrayList<>()
-        : null);
     
+    this.writeContexts.removeIf(WriteContext::isNotValid);
```
this is here for GC.  I don't think we should call it a ton because it is probably expensive if the list gets long.  Anything that goes invalid before being removed will just get dropped on the floor because prepareToSend returns false when there the entire group is unable to commit.  maybe we only call this when fail fast is true and unable to acquire the lock?

myronkscott commented:
I think I addressed most things.  It is better, thank you.  We should discuss tc.properties and removeIf next week because I don't think I complete understood.
